pkg/geo/geomfn/mvtgeom.go

```diff
@@ -0,0 +1,351 @@
+// Since snapping to the grid can create an invalid geometry and making it valid can create float values,
+// we iterate up to 3 times to generate a valid geometry with integer coordinates.
+func snapToGridAndValidateBasicPolygon(
+	g geo.Geometry, basicType geopb.ShapeType,
+) (geo.Geometry, error) {
+	iterations := 0
+	const maxIterations = 3
+	valid := false
+
+	var err error
+	g, err = snapToIntegersGrid(g)
+	if err != nil {
+		return geo.Geometry{}, err
+	}
+	valid, err = IsValid(g)
+	if err != nil {
+		return geo.Geometry{}, errors.Wrap(err, "failed to check validity")
+	}
+	for !valid && iterations < maxIterations {
+		g, err = MakeValid(g)
+		if err != nil {
+			return geo.Geometry{}, errors.Wrap(err, "failed to make valid")
+		}
+
+		g, err = snapToIntegersGrid(g)
+		if err != nil {
+			return geo.Geometry{}, err
+		}
+		converted, err := convertToBasicType(g, basicType)
+		if err != nil {
+			return geo.Geometry{}, errors.Wrap(err, "failed to convert to basic type")
+		}
+		g = converted
+		valid, err = IsValid(g)
+		if err != nil {
+			return geo.Geometry{}, errors.Wrap(err, "failed to check validity")
+		}
+		iterations++
+	}
+	if !valid {
+		return geo.Geometry{}, nil
+	}
+
+	gt, err := g.AsGeomT()
+	if err != nil {
+		return geo.Geometry{}, errors.Wrap(err, "failed to convert to geom.T")
+	}
+	forceClockwiseOrientation(gt)
+	g, err = geo.MakeGeometryFromGeomT(gt)
+	if err != nil {
+		return geo.Geometry{}, errors.Wrap(err, "failed to make geometry from geom.T")
+	}
+	return g, nil
+}
+
+func snapToIntegersGrid(g geo.Geometry) (geo.Geometry, error) {
+	offset := geom.Coord{0, 0, 0, 0}
+	grid := geom.Coord{1, 1, 0, 0}
+	g, err := SnapToGrid(g, offset, grid)
+	if err != nil {
+		return geo.Geometry{}, errors.Wrap(err, "failed to snap to grid")
+	}
+	return g, nil
+}
+
+// forceClockwiseOrientation reverses the Polygon's rings that are CCW.
+// Works for MultiPolygons and GeometryCollections too.
+func forceClockwiseOrientation(gt geom.T) {
+	switch gt := gt.(type) {
+	case *geom.Polygon:
+		if gt.Empty() {
+			return
+		}
+		externalRing := gt.LinearRing(0)
+		if isCCW(externalRing) {
+			externalRing.Reverse()
+		}
+		for i := 1; i < gt.NumLinearRings(); i++ {
+			internalRing := gt.LinearRing(i)
+			if !isCCW(internalRing) {
+				internalRing.Reverse()
+			}
+		}
+	case *geom.MultiPolygon:
+		for i := 0; i < gt.NumPolygons(); i++ {
+			forceClockwiseOrientation(gt.Polygon(i))
+		}
+	case *geom.GeometryCollection:
+		for i := 0; i < gt.NumGeoms(); i++ {
+			forceClockwiseOrientation(gt.Geom(i))
+		}
+	}
+}
+
+type areaCalculatable interface {
+	Area() float64
+}
+
+func isCCW(geometry areaCalculatable) bool {
+	return geometry.Area() <= 0
+}
+
+func overwriteMinusZero(val float64) float64 {
+	if val == -0 {
+		return 0
+	}
+	return val
+}
+
+// For a given geometry, look for the highest dimensional basic type: point, line or polygon.
+func getBasicType(gt geom.T) (geopb.ShapeType, error) {
```

Comment:
can we use `shapeTypeFromGeomT`, and only do the recursive step for `GeometryCollection`?

Reply:
Hmm `shapeTypeFromGeomT` can return `MultiLineString`, `MultiPolygon` or a `GeometryCollection`, while `getBasicType` is supposed to return only `Point`, `Line` or a `Polygon`. Not sure if we can use it here